### PR TITLE
Feature: Cog's can now have external config

### LINKED ISSUE #129: Add a way for bot extensions to have an external type of configuration
### Link

https://github.com/Rapptz/discord.py/projects/3#card-58810575

### PR DESCRIPTION
**Pull request by Skelmis**
## Summary

It resolves #129 by implementing the ability to have keyword arguments passed through

See the thread under nc-development for testing

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [x] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: nextcord/__init__.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 __author__ = 'tag-epic & Rapptz'
 __license__ = 'MIT'
 __copyright__ = 'Copyright 2015-present Rapptz'
-__version__ = '2.0.0a3'
+__version__ = '2.0.0a4'
```
Comment (ooliver1) [inline nextcord/__init__.py:16]: the versionadded was below ``components`` as that was added in 2.0, could you move it like this?

Comment (ooliver1) [inline nextcord/__init__.py:16]: i assume you asked epic about version bumping?

File: nextcord/ext/commands/bot.py [hunk 2]
```diff
@@ -707,6 +733,11 @@ def load_extension(self, name: str, *, package: Optional[str] = None) -> None:
             Defaults to ``None``.
 
             .. versionadded:: 1.7
+        extras: Optional[:class:`dict`]
+            A mapping of kwargs to values to be passed to your
+            cog's ``__init__`` method as key word arguments.
+
+            .. versionadded:: 2.0.0a4
```
Comment (ooliver1) [inline nextcord/ext/commands/bot.py:740]: a4 should not be added as thats ``alpha`` and ``serial 4``

File: nextcord/ext/commands/errors.py [hunk 3]
```diff
@@ -859,6 +859,14 @@ class NoEntryPointError(ExtensionError):
     def __init__(self, name: str) -> None:
         super().__init__(f"Extension {name!r} has no 'setup' function.", name=name)
 
+class InvalidSetupArguments(ExtensionError):
+    """An exception raised when an extension contains a ``setup`` function which does except ``kwargs`` but ``kwargs`` were passed.
```
Comment (ooliver1) [inline nextcord/ext/commands/errors.py:863]: Is it at all possible to wrap this to two lines as its 132 characters long? Simply wrapping it in the middle should work.

File: nextcord/ext/commands/errors.py [hunk 4]
```diff
@@ -859,6 +859,14 @@ class NoEntryPointError(ExtensionError):
     def __init__(self, name: str) -> None:
         super().__init__(f"Extension {name!r} has no 'setup' function.", name=name)
 
+class InvalidSetupArguments(ExtensionError):
```
Comment (ooliver1) [inline nextcord/ext/commands/errors.py:862]: you havent added this to ``docs/ext/commands/api.rst``?

```rst
.. autoexception:: nextcord.ext.commands.InvalidSetupArguments
    :members:
```
then ``.. exception_hierarchy::`` should contain it too (no need to mention "This inherits from :exc:`ExtensionError`" as the exception hierarchy shows this)

**Comment by ooliver1** [comment]
forgot to set as a full review sorry

File: nextcord/__init__.py (hunk 1 above)
Comment (TAG-Epic) [reply nextcord/__init__.py:16]: Don't bump the version

Comment (Skelmis) [reply nextcord/__init__.py:16]: So versionadded will be a3 then in docs?

File: nextcord/ext/commands/bot.py (hunk 2 above)
Comment (Skelmis) [reply nextcord/ext/commands/bot.py:740]: Yep. What should it be then? Just 2.0.0

File: nextcord/ext/commands/errors.py (hunk 4 above)
Comment (Skelmis) [reply nextcord/ext/commands/errors.py:862]: In relation to
>  (no need to mention This inherits from :exc:`ExtensionError` as the exception hierarchy shows this)
All other errors have it

File: nextcord/__init__.py (hunk 1 above)
Comment (ooliver1) [reply nextcord/__init__.py]: No it should just be 2.0.0, as the ``a`` means alpha and it's not a proper release

File: nextcord/ext/commands/bot.py [hunk 5]
```diff
@@ -669,8 +674,18 @@ def _load_from_module_spec(self, spec: importlib.machinery.ModuleSpec, key: str)
             del sys.modules[key]
             raise errors.NoEntryPointError(key)
 
+        params = inspect.signature(setup).parameters
+        has_kwargs = len(params) > 1
+
+        if extras:
```
Comment (dzshn) [inline nextcord/ext/commands/bot.py:680]: Someone could provide a falsy object and `TypeError` wouldn't be raised, didn't notice this earlier
```suggestion
        if extras is not None:
```

**Review by dzshn — APPROVED** [review]

**Review by DenverCoder1 — APPROVED** [review]

**Review by VincentRPS — APPROVED** [review]

File: nextcord/ext/commands/bot.py [hunk 6]
```diff
@@ -653,7 +653,12 @@ def _call_module_finalizers(self, lib: types.ModuleType, key: str) -> None:
                 if _is_submodule(name, module):
                     del sys.modules[module]
 
-    def _load_from_module_spec(self, spec: importlib.machinery.ModuleSpec, key: str) -> None:
+    def _load_from_module_spec(
+            self,
+            spec: importlib.machinery.ModuleSpec,
+            key: str,
+            extras: Optional[Dict[str, Any]] = None
```
Comment (veni-vidi-code) [inline nextcord/ext/commands/bot.py:660]: Maybe use Missing instead?

File: nextcord/ext/commands/bot.py [hunk 7]
```diff
@@ -685,7 +700,13 @@ def _resolve_name(self, name: str, package: Optional[str]) -> str:
         except ImportError:
             raise errors.ExtensionNotFound(name)
 
-    def load_extension(self, name: str, *, package: Optional[str] = None) -> None:
+    def load_extension(
+            self,
+            name: str,
+            *,
+            package: Optional[str] = None,
+            extras: Optional[Dict[str, Any]] = None
```
Comment (veni-vidi-code) [inline nextcord/ext/commands/bot.py:708]: same here

File: nextcord/ext/commands/bot.py (hunk 6 above)
Comment (Skelmis) [reply nextcord/ext/commands/bot.py:660]: Based on epics thumbs down imma resolve this

**Review by TAG-Epic — CHANGES_REQUESTED** [review]

**Comment by Skelmis** [comment]
Just awaiting approval

@TAG-Epic @ooliver1

**Review by TAG-Epic — APPROVED** [review]

**Review by ooliver1 — APPROVED** [review]

**Comment by sairam4123** [comment]
How does it work?

**Comment by Skelmis** [comment]
> How does it work?

You pass in a dict to the extras argument and those values will get called in the cogs setup method as arguments

**Comment by ooliver1** [comment]
that dict is like kwargs passed to setup() that you can pass to the cog if you want

**Pull request merged**